cmd/soroban-rpc/internal/db/transaction.go

```diff
@@ -0,0 +1,203 @@
+	ApplicationOrder int32
+	Successful       bool
+	Ledger           ledgerbucketwindow.LedgerInfo
+}
+
+type TransactionHandler struct {
+	stmtCache  *sq.StmtCache
+	db         db.SessionInterface
+	passphrase string
+}
+
+func NewTransactionHandler(db db.SessionInterface, passphrase string) *TransactionHandler {
+	return &TransactionHandler{db: db, passphrase: passphrase}
+}
+
+func (txn *TransactionHandler) InsertTransactions(lcm xdr.LedgerCloseMeta) error {
+	reader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(txn.passphrase, lcm)
+	if err != nil {
+		return err
+	}
+	txCount := lcm.CountTransactions()
+	transactions := make(map[xdr.Hash]ingest.LedgerTransaction, txCount)
+
+	for i := 0; i < txCount; i++ {
+		tx, err := reader.Read()
+		if err != nil {
+			return err
+		}
+
+		if tx.Envelope.IsFeeBump() {
+			transactions[tx.Result.InnerHash()] = tx
+		}
+		transactions[tx.Result.TransactionHash] = tx
+	}
+
+	query := sq.
+		Insert(transactionTableName).
+		RunWith(txn.stmtCache).
+		Columns("hash", "ledger_sequence", "application_order")
+	for hash, tx := range transactions {
+		query = query.Values(hash, lcm.LedgerSequence(), tx.Index)
+	}
+
+	_, err = query.Exec()
+	return err
+}
+
+func (txn *TransactionHandler) GetLedgerRange() ledgerbucketwindow.LedgerRange {
+	var ledgerRange ledgerbucketwindow.LedgerRange
+	newestQ := sq.
+		Select("meta").
+		From(ledgerCloseMetaTableName).
+		OrderBy("sequence DESC").
+		Limit(1).
+		RunWith(txn.stmtCache).
+		QueryRow()
+	oldestQ := sq.
+		Select("meta").
+		From(ledgerCloseMetaTableName).
+		OrderBy("sequence ASC").
+		Limit(1).
+		RunWith(txn.stmtCache).
+		QueryRow()
+
+	var row1, row2 []byte
+	if err := newestQ.Scan(&row1); err != nil {
+		return ledgerRange
+	}
+	if err := oldestQ.Scan(&row2); err != nil {
+		return ledgerRange
+	}
+
+	var lcm1, lcm2 xdr.LedgerCloseMeta
+	if err := lcm1.UnmarshalBinary(row1); err != nil {
+		return ledgerRange
+	}
+	if err := lcm2.UnmarshalBinary(row2); err != nil {
+		return ledgerRange
+	}
+
+	return ledgerbucketwindow.LedgerRange{
+		FirstLedger: ledgerbucketwindow.LedgerInfo{
+			Sequence:  lcm1.LedgerSequence(),
+			CloseTime: int64(lcm1.LedgerHeaderHistoryEntry().Header.ScpValue.CloseTime),
+		},
+		LastLedger: ledgerbucketwindow.LedgerInfo{
+			Sequence:  lcm2.LedgerSequence(),
+			CloseTime: int64(lcm2.LedgerHeaderHistoryEntry().Header.ScpValue.CloseTime),
+		},
+	}
+}
+
+func (txn *TransactionHandler) GetTransactionByHash(hash string) (
+	xdr.LedgerCloseMeta, ingest.LedgerTransaction, error,
+) {
+	rawHash, err := hex.DecodeString(hash)
+	if err != nil {
+		return xdr.LedgerCloseMeta{}, ingest.LedgerTransaction{}, err
+	}
+
+	rows := sq.
+		Select("t.application_order", "lcm.meta").
+		From(fmt.Sprintf("%s t", transactionTableName)).
+		Join(fmt.Sprintf("%s lcm ON (t.ledger_sequence = lcm.sequence)", ledgerCloseMetaTableName)).
+		Where(sq.Eq{"hash": rawHash}).
+		Limit(1).
+		RunWith(txn.stmtCache).
+		QueryRow()
```

Comment:
@sydneynotthecity @chowbao is this a dumb way to do this? I think the `JOIN` is inevitable. Also, would `USING (sequence)` be superior in any way?

Reply:
`USING (sequence)` is not superior in any way. Some argue it's easier to read, but the SQL interpreter will resolve the two the exact same way.

The biggest performance gain you'll get for this query is to confirm that the first listed table is indexed by your join key. ie, does the `transactions` table have a index on `sequence`? or does the `lcm`? Whichever table contains that index needs to come first.

Reply:
There are implicit indexes on primary keys, but not foreign keys.

I'm not super familiar with the sqlite query planner--you may be right that the `WHERE` clause kicks in first....but it could also be the `JOIN` 🤷‍♀️ I think it might be worthwhile to just run an `EXPLAIN ANALYZE` against a join version of the query vs subquery to see if there is any performance differences. My guess is the differences will be negligible. If you do change the query, I would opt for a subquery instead of two queries. The query planner will resolve as two queries anyway.

Reply:
Good call, thanks! It looks like it does the smart thing for this version, anyway:

```
EXPLAIN QUERY PLAN 
  SELECT t.application_order, lcm.meta 
  FROM transactions t JOIN ledger_close_meta lcm ON (t.ledger_sequence = lcm.sequence) 
  WHERE t.hash = '29994b6f84a0fcd19b5bc5f35e692f418feb4ea3ef9a2edf71c61c4f217e079d'
  LIMIT 1;

QUERY PLAN
|--SEARCH t USING INDEX sqlite_autoindex_transactions_1 (hash=?)
`--SEARCH lcm USING INTEGER PRIMARY KEY (rowid=?)
```

which tells me, if I'm reading it right, that it does the `WHERE` first using the pkey index, then uses the pkey index on the `ledger_sequence` from that result row :ok_hand: which is exactly what I expanded it to anyway.

Yay for smort planners. Thanks again!